// polymer-analyzer/lib/polymer/js-utils.js
"use strict";
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
Object.defineProperty(exports, "__esModule", { value: true });
const doctrine = require("doctrine");
const escodegen = require("escodegen");
const esutil_1 = require("../javascript/esutil");
const jsdoc = require("../javascript/jsdoc");
const model_1 = require("../model/model");
/**
 * Create a ScannedProperty object from an estree Property AST node.
 */
function toScannedPolymerProperty(node, sourceRange, document) {
    const parsedJsdoc = jsdoc.parseJsdoc(esutil_1.getAttachedComment(node) || '');
    const description = parsedJsdoc.description.trim();
    const maybeName = esutil_1.objectKeyToString(node.key);
    const warnings = [];
    if (!maybeName) {
        warnings.push(new model_1.Warning({
            code: 'unknown-prop-name',
            message: `Could not determine name of property from expression of type: ` +
                `${node.key.type}`,
            sourceRange: sourceRange,
            severity: model_1.Severity.WARNING,
            parsedDocument: document
        }));
    }
    let type = esutil_1.closureType(node.value, sourceRange, document);
    const typeTag = jsdoc.getTag(parsedJsdoc, 'type');
    if (typeTag) {
        type = doctrine.type.stringify(typeTag.type) || type;
    }
    const name = maybeName || '';
    const result = {
        name,
        type,
        description,
        sourceRange,
        warnings,
        astNode: node,
        isConfiguration: configurationProperties.has(name),
        jsdoc: parsedJsdoc,
        privacy: getOrInferPrivacy(name, parsedJsdoc)
    };
    return result;
}
exports.toScannedPolymerProperty = toScannedPolymerProperty;
;
/**
 * Properties on Polymer element prototypes that are part of Polymer's
 * configuration syntax.
 */
const configurationProperties = new Set([
    'attached',
    'attributeChanged',
    'beforeRegister',
    'configure',
    'constructor',
    'created',
    'detached',
    'enableCustomStyleProperties',
    'extends',
    'hostAttributes',
    'is',
    'listeners',
    'mixins',
    'observers',
    'properties',
    'ready',
    'registered',
]);
/**
 * Create a ScannedMethod object from an estree Property AST node.
 */
function toScannedMethod(node, sourceRange, document) {
    const scannedMethod = toScannedPolymerProperty(node, sourceRange, document);
    if (scannedMethod.type === 'Function' ||
        scannedMethod.type === 'ArrowFunction') {
        const value = node.value;
        const paramTags = new Map();
        if (scannedMethod.jsdoc) {
            for (const tag of (scannedMethod.jsdoc.tags || [])) {
                if (tag.title === 'param' && tag.name) {
                    paramTags.set(tag.name, tag);
                }
                else if (tag.title === 'return' || tag.title === 'returns') {
                    scannedMethod.return = {};
                    if (tag.type) {
                        scannedMethod.return.type = doctrine.type.stringify(tag.type);
                    }
                    if (tag.description) {
                        scannedMethod.return.desc = tag.description;
                    }
                }
            }
        }
        scannedMethod.params = (value.params || []).map((nodeParam) => {
            let type = undefined;
            let description = undefined;
            // With ES6 we can have a lot of param patterns. Best to leave the
            // formatting to escodegen.
            const name = escodegen.generate(nodeParam);
            const tag = paramTags.get(name);
            if (tag) {
                if (tag.type) {
                    type = doctrine.type.stringify(tag.type);
                }
                if (tag.description) {
                    description = tag.description;
                }
            }
            return { name, type, description };
        });
    }
    return scannedMethod;
}
exports.toScannedMethod = toScannedMethod;
function getOrInferPrivacy(name, annotation, defaultPrivacy = 'public') {
    const explicitPrivacy = jsdoc.getPrivacy(annotation);
    const specificName = name.slice(name.lastIndexOf('.') + 1);
    if (explicitPrivacy) {
        return explicitPrivacy;
    }
    if (specificName.startsWith('__')) {
        return 'private';
    }
    else if (specificName.startsWith('_')) {
        return 'protected';
    }
    else if (specificName.endsWith('_')) {
        return 'private';
    }
    else if (configurationProperties.has(specificName)) {
        return 'protected';
    }
    return defaultPrivacy;
}
exports.getOrInferPrivacy = getOrInferPrivacy;
