// runjs/lib/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.logger = undefined;
exports.RunJSError = RunJSError;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Needed to use ES5 inheritance, because of issues with Error subclassing for Babel
function RunJSError(message) {
  this.name = 'RunJSError';
  this.message = message && message.split('\n')[0]; // assign only first line
}
RunJSError.prototype = Object.create(Error.prototype);
RunJSError.prototype.constructor = RunJSError;

var logger = exports.logger = {
  debug: function debug() {
    console.log(_chalk2.default.blue.apply(_chalk2.default, arguments));
  },
  info: function info() {
    console.log(_chalk2.default.bold.apply(_chalk2.default, arguments));
  },
  log: function log() {
    var _console;

    (_console = console).log.apply(_console, arguments);
  },
  warning: function warning() {
    console.warn(_chalk2.default.yellow.apply(_chalk2.default, arguments));
  },
  error: function error() {
    console.error(_chalk2.default.red.apply(_chalk2.default, arguments));
  }
};
